File: zadani3/zadanie3.cpp
#include <iostream>
#include <vector>
#include <cstdint>

using namespace std;

int64_t gcd(int64_t a, int64_t b);
vector<int64_t> euler(int a);
int64_t polyEval(const vector<int64_t>& coeff, int b);
int64_t denPow(int b, int a);
pair<int64_t, int64_t> sum(int a, int b);

int main() {
    int a, b;
    cin >> a >> b;
    if (b == 1) {
        cout << "infinity\n";
    } else{
        pair<int64_t, int64_t> result = sum(a, b);
        cout << result.first << "/" << result.second << endl;
    }
    return 0;
}

int64_t gcd(int64_t a, int64_t b) { 
    while (b != 0) {
        int64_t temp = a % b;
        a = b;
        b = temp;
    }
    return a;
}

vector<int64_t> euler(int a) { 
    vector<int64_t> prev = {1};
    for (int n = 2; n <= a; ++n) {
        vector<int64_t> cur(n, 1);
        for (int k = 1; k < n - 1; ++k){
            cur[k] = (k + 1) * prev[k] + (n - k) * prev[k - 1]; 
        }
        prev.swap(cur);
    }
    return prev;
}

int64_t polyEval(const vector<int64_t>& coef, int b) { 
    int64_t res = 0;
    int64_t powB = 1;
    for (size_t k = 0; k < coef.size(); ++k) {
        res += coef[k] * powB;
        powB *= b;
    }
    return res;
}

int64_t denPow(int b, int a) { 
    int64_t res = 1;
    for (int i = 0; i <= a; ++i) {
        res *= (b - 1);
    }
    return res;
}

pair<int64_t, int64_t> sum(int a, int b) { 
    vector<int64_t> coeff = euler(a);
    int64_t poly = polyEval(coeff, b);
    int64_t num = b * poly;
    int64_t den = denPow(b, a);
    int64_t g = gcd(num, den);
    return {num / g, den / g};
}

File: zadanie3.go
package main

import "fmt"

func main() {
	var a, b int
	fmt.Scan(&a, &b)
	if b == 1 {
		fmt.Println("infinity")
	} else {
		num, den := sum(a, b)
		fmt.Printf("%d/%d\n", num, den)
	}
}

func gcd(a, b int64) int64 {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func euler(a int) []int64 {
	prev := []int64{1}
	for n := 2; n <= a; n++ {
		cur := make([]int64, n)
		cur[0] = 1
		cur[n-1] = 1
		for k := 1; k < n-1; k++ {
			cur[k] = int64(k+1)*prev[k] + int64(n-k)*prev[k-1]
		}
		prev = cur
	}
	return prev
}

func polyEval(coef []int64, b int) int64 {
	var res int64 = 0
	powB := int64(1)
	for k := 0; k < len(coef); k++ {
		res += coef[k] * powB
		powB *= int64(b)
	}
	return res
}

func denPow(b, a int) int64 {
	d := int64(1)
	for i := 0; i <= a; i++ {
		d *= int64(b - 1)
	}
	return d
}

func sum(a, b int) (int64, int64) {
	coeff := euler(a)
	poly := polyEval(coeff, b)
	num := int64(b) * poly
	den := denPow(b, a)
	g := gcd(num, den)
	return num / g, den / g
}
